code/datums/components/food/decomposition.dm

```diff
@@ -0,0 +1,162 @@
+//"Don't leave food on the floor, that's how we get ants"
+
+#define DECOMPOSITION_TIME 10 MINUTES
+#define DECOMPOSITION_TIME_RAW 5 MINUTES
+#define DECOMPOSITION_TIME_GROSS 7 MINUTES
+
+/datum/component/decomposition
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	/// Makes sure food only starts decomposing if a player's EVER picked it up before
+	var/handled = FALSE
+	/// Used to stop food in someone's hand & in storage slots from decomposing.
+	var/protected = FALSE
+	/// Used to stop the timer & check for the examine proc
+	var/timerid
+	/// Used so the timer won't reset.
+	var/time_remaining = DECOMPOSITION_TIME
+	/// Used to give raw/gross food lower timers
+	var/decomp_flags
+	/// Used for examining
+	var/examine_type = 0
+
+/datum/component/decomposition/Initialize(decomp_flags = NONE)
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.decomp_flags = decomp_flags
+
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/handle_movement)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP, //person picks up an item
+		COMSIG_STORAGE_ENTERED), //Object enters a storage object (boxes, etc.)
+		.proc/picked_up)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_DROPPED, //Object is dropped anywhere
+		COMSIG_STORAGE_EXITED), //Object exits a storage object (boxes, etc)
+		.proc/dropped)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine) // Self-explanitory
+
+	if(decomp_flags & GROSS && !(decomp_flags & RAW))
+		time_remaining = DECOMPOSITION_TIME_GROSS
+		examine_type = 1
+	if(decomp_flags & RAW) // Raw food overrides gross
+		time_remaining = DECOMPOSITION_TIME_RAW
+		examine_type = 2
+
+/datum/component/decomposition/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP,
+		COMSIG_STORAGE_ENTERED,
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_ITEM_DROPPED,
+		COMSIG_STORAGE_EXITED,
+		COMSIG_PARENT_EXAMINE))
+
+/datum/component/decomposition/proc/handle_movement() // Assuming it is guaranteed to be food
+	var/obj/item/food/food = parent
+	var/atom/last_loc = food.loc
+
+	var/clean = FALSE // Used to check if it's on a clean surface
+	while(last_loc && !isarea(last_loc))
+		if(HAS_TRAIT(food.loc, TRAIT_PROTECT_FOOD))
+			clean = TRUE
+			break
+		else if(locate(/obj/structure/table) in food.loc)
+			clean = TRUE
+			break
+		last_loc = last_loc.loc
```

Comment:
Does this work? you're modifying last_loc, but you only ever check last_loc
Speaking of which, should this really apply to stuff that's inside a container like a doughnut box or a bag? (This isn't a rhetorical question, I want to know what you think/why)

Reply:
It functions, yea.
When you ask if it should apply to stuff inside a container, what do you mean? Like, should food be preserved when inside boxes?? If you're asking that, yes, in my mind the main reason food is decomposing is because the ants are getting to it, and ants can't get into sealed containers.

Reply:
You're doing these checks each time the loop iterates
```
if(HAS_TRAIT(food.loc, TRAIT_PROTECT_FOOD))
	clean = TRUE
	break
else if(locate(/obj/structure/table) in food.loc)
	clean = TRUE
	break
```
but nearest I can tell the food's location isn't changing inside the loop. It feels like you want to use last_loc here, but I'm not really sure

Reply:
ah right, I should've made a check to make sure it doesn't do anything `if(food.loc == last_loc)`, that should reduce the times it checks if it's not actually changing locations

Reply:
Yea I'm unsure how to make that function properly. 
I did this:
```
/datum/component/decomposition/proc/handle_movement() // Assuming it is guaranteed to be food
	var/obj/food = parent // Doesn't HAVE to be food, that's just what it's intended for
	var/atom/loc_atom = food.loc
	if(food.loc == last_loc)
		return
	var/clean = FALSE // Used to check if it's on a clean surface
	var/anti_loop = FALSE
	while(loc_atom && !isarea(loc_atom))
		if(anti_loop)
			return
		if(!(istype(food.loc, /turf/open)))
			clean = TRUE
			break
		else if(locate(/obj/structure/table) in food.loc)
			clean = TRUE
			break
		else if(locate(/obj/machinery/conveyor) in food.loc)
			clean = TRUE
			break
		anti_loop = TRUE
		loc_atom = loc_atom.loc
		last_loc = food.loc
```
Moving last_loc to be its own variable. While it does stop the proc from firing twice in a row like it should, for some reason that makes it so the food doesn't start its decomposition timer now.

Reply:
Yep I haven't the slightest clue how to make it not check the locations for each time in the loop.

Reply:
you should talk to kyler, or bug me on discord. it shouldn't be needed if you only want to check for a table/table like thing

Reply:
Fixed it! Just completely removed the while loop, it works just the same.